packages/compass-aggregations/src/modules/collections-fields.ts

```diff
@@ -0,0 +1,204 @@
+  fields: string[];
+  error?: Error;
+};
+
+type State = Record<string, CollectionData>;
+
+export const INITIAL_STATE: State = {};
+
+export default function reducer(
+  state = INITIAL_STATE,
+  action: AnyAction
+): State {
+  if (
+    isAction<CollectionsFetchedAction>(action, ActionTypes.CollectionsFetch)
+  ) {
+    return action.data;
+  }
+
+  if (
+    isAction<CollectionDataUpdatedAction>(
+      action,
+      ActionTypes.CollectionDataUpdated
+    )
+  ) {
+    return {
+      ...state,
+      [action.collection]: action.data,
+    };
+  }
+
+  if (
+    isAction<CollectionFieldsFetchedAction>(
+      action,
+      ActionTypes.CollectionFieldsFetched
+    )
+  ) {
+    return {
+      ...state,
+      [action.collection]: {
+        isLoading: false,
+        fields: action.fields,
+        type: action.collectionType,
+      },
+    };
+  }
+
+  return state;
+}
+
+export const setCollections = (
+  collections: CollectionInfo[]
+): PipelineBuilderThunkAction<void, CollectionsFetchedAction> => {
+  return (dispatch, getState) => {
+    const currentData = getState().collectionsFields;
+
+    const newData: State = {};
+    collections.forEach(({ name, type }) => {
+      newData[name] = currentData[name] ?? {
+        fields: [],
+        isLoading: false,
+        type,
+      };
+    });
+
+    if (isEqual(Object.keys(currentData), Object.keys(newData))) {
+      return;
+    }
+
+    dispatch({
+      type: ActionTypes.CollectionsFetch,
+      data: newData,
+    });
+  };
+};
+
+export const setCollectionFields = (
+  collection: string,
+  collectionType: CollectionType,
+  fields: string[]
+): CollectionFieldsFetchedAction => ({
+  type: ActionTypes.CollectionFieldsFetched,
+  collection,
+  fields,
+  collectionType,
+});
+
+export const fetchCollectionFields = (
+  collection: string
+): PipelineBuilderThunkAction<Promise<void>, CollectionDataUpdatedAction> => {
+  return async (dispatch, getState) => {
+    const {
+      collectionsFields,
+      namespace,
+      dataService: { dataService },
+    } = getState();
+
+    if (!dataService) {
+      return;
+    }
+
+    const collectionInfo = collectionsFields[collection];
+    if (collectionInfo && collectionInfo.fields.length > 0) {
+      return;
+    }
+
+    dispatch({
+      type: ActionTypes.CollectionDataUpdated,
+      collection,
+      data: {
+        ...collectionInfo,
+        fields: [],
+        isLoading: true,
+        error: undefined,
+      },
+    });
+
+    try {
+      const { database } = toNS(namespace);
+      const namespaceToQuery = `${database}.${collection}`;
+
+      const documents =
+        collectionInfo.type === 'collection'
+          ? await dataService.find(
+              namespaceToQuery,
+              {},
+              {
+                sort: { $natural: -1 },
+                limit: 1,
+                maxTimeMS: FETCH_SCHEMA_MAX_TIME_MS,
+              }
+            )
+          : await dataService.sample(
```

Comment:
Something I'm not sure on is `$sample` performance with time series collections and views. We want to avoid a collection scan if possible (collection scans happen when running count operations on those). Does it perform well?

Reply:
We did add timeout there, but it its definitely not the best way to fetch a single document to figure out the schema. 

With collections, we are using find and it just works fine.

Reply:
Ah yup all good.